Code/src/main/java/nl/utwente/group10/haskell/expr/Apply.java

```diff
@@ -1,66 +1,60 @@
 package nl.utwente.group10.haskell.expr;
 
+import nl.utwente.group10.haskell.env.Env;
 import nl.utwente.group10.haskell.exceptions.HaskellTypeError;
+import nl.utwente.group10.haskell.hindley.GenSet;
+import nl.utwente.group10.haskell.hindley.HindleyMilner;
 import nl.utwente.group10.haskell.type.FuncT;
 import nl.utwente.group10.haskell.type.Type;
 
-import java.util.Arrays;
-
 /**
- * A type of expression that is a lazy application of a function with certain arguments.
+ * Lazy application of an argument to a function.
  */
-public class Apply extends Func {
+public class Apply extends Expr {
     /**
-     * The expression to apply arguments to.
+     * The expression to apply the argument to.
      */
-    private final Func func;
+    private final Expr func;
 
     /**
-     * The arguments to apply.
+     * The argument to apply.
      */
-    private final Expr[] args;
+    private final Expr arg;
 
     /**
-     * Constructs a new function application object. The resulting type will be calculated based on the given arguments.
-     * The type of the arguments is validated, and an exception is thrown when one of the arguments is incorrect.
+     * Applies an argument to a function to produce a new expression. The application is lazy, the type needs to be
+     * analyzed before it can be resolved.
```

Comment:
If he/she feels like it. I intended it as a warning - one might want to chain multiple applies first and then call ```analyze``` (or just ignore to do it). Their choice, however, users of any program implementing this library may not be happy with that decision.